File: python/tests/test_ad_star.py
"""AD* (Anytime Dynamic A*): the executed trajectory is belief-optimal like D*
Lite (equals A* when the freespace assumption doesn't mislead), replans a valid
detour when an obstacle is revealed, handles the no-path case, validates params,
and — combining ARA*'s anytime repair — emits a strictly suboptimal first
solution that improves to the optimum as ε -> 1."""

from __future__ import annotations

import io
import json
from pathlib import Path

import pytest
from conftest import REPO_ROOT, config, grid_from, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell
from navigation.global_planning.search.ad_star import ADStar
from navigation.global_planning.search.astar import AStar
from navigation.maps.loader import load_map, load_scenario
from navigation.maps.occupancy_grid import OccupancyGrid2D

# Same field as the ARA* anytime fixture but with start/goal swapped so the two
# obstacles sit near the *robot*: AD*'s backward (goal->start) search rushes toward
# the robot under the inflated key and rounds them the costly way first, then
# repairs to the optimum. (ARA*'s forward search wanted them near the goal instead.)
_FIELD_ROWS = [
    ".............",
    "..........#..",
    "...........#.",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
    ".............",
]
_FIELD_START: Cell = (0, 11)
_FIELD_GOAL: Cell = (11, 12)


def _path_cost(grid: OccupancyGrid2D, path: list[Cell]) -> float:
    total = 0.0
    for a, b in zip(path, path[1:], strict=False):
        for cell, cost in grid.neighbors(a):
            if cell == b:
                total += cost
                break
    return total


def _is_valid_trajectory(grid: OccupancyGrid2D, path: list[Cell]) -> bool:
    # Every consecutive pair is a legal move under GROUND TRUTH (grid.neighbors),
    # proving the executed trajectory never crosses a real obstacle.
    for a, b in zip(path, path[1:], strict=False):
        if not any(cell == b for cell, _ in grid.neighbors(a)):
            return False
    return True


def _anytime_config(
    tmp_path: Path, eps_start: float, eps_step: float, sensor_radius: int
) -> ParamSet:
    return ParamSet.from_yaml(
        write_config(
            tmp_path / "ad_any.yaml",
            "ad_star",
            [
                {"name": "eps_start", "type": "float", "default": eps_start,
                 "min": 1.0, "max": 10.0, "description": "start eps"},
                {"name": "eps_final", "type": "float", "default": 1.0,
                 "min": 1.0, "max": 10.0, "description": "final eps"},
                {"name": "eps_step", "type": "float", "default": eps_step,
                 "min": 0.01, "max": 10.0, "description": "eps decrement"},
                {"name": "sensor_radius", "type": "int", "default": sensor_radius,
                 "min": 1, "max": 50, "description": "sensor radius"},
                {"name": "max_expansions", "type": "int", "default": 100000,
                 "min": 1, "max": 100000000, "description": "cap"},
            ],
        )
    )


def test_ad_star_static_trajectory_matches_astar_optimum_on_maze01() -> None:
    # eps_final=1 makes each belief-optimal step follow the true shortest path; on
    # maze01 the freespace assumption never pushes the robot off it, so the executed
    # trajectory equals the omniscient A* optimum (the D* Lite guarantee under AD*).
    grid = load_map(str(REPO_ROOT / "maps/grid/maze01.yaml"), seed=0, connectivity=8)
    sc = load_scenario(str(REPO_ROOT / "maps/scenarios/maze01_s1.yaml"))
    start = grid.world_to_cell(*sc.start)
    goal = grid.world_to_cell(*sc.goal)
    res = ADStar(config("ad_star")).plan(grid, start, goal)
    ast = AStar(config("astar")).plan(grid, start, goal)
    assert res.success and ast.success
    assert res.path[0] == start and res.path[-1] == goal
    assert _is_valid_trajectory(grid, res.path)
    assert res.cost == pytest.approx(ast.cost)


def test_ad_star_replans_valid_detour_when_trap_revealed() -> None:
    # dstar_trap01 is a C-shaped trap whose mouth faces the robot. Assuming freespace,
    # AD* heads straight in, senses the back wall, and must replan out and around. The
    # executed trajectory stays collision-free in ground truth, reaches the goal, and —
    # because the trap forced a real detour — is strictly longer than the omniscient
    # A* optimum. At least one replan must have fired.
    grid = load_map(str(REPO_ROOT / "maps/grid/dstar_trap01.yaml"), seed=0, connectivity=8)
    sc = load_scenario(str(REPO_ROOT / "maps/scenarios/dstar_trap01_s1.yaml"))
    start = grid.world_to_cell(*sc.start)
    goal = grid.world_to_cell(*sc.goal)
    res = ADStar(config("ad_star")).plan(grid, start, goal)
    ast = AStar(config("astar")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert _is_valid_trajectory(grid, res.path)
    assert res.stats.iterations >= 1  # replanned at least once after sensing the trap
    assert res.cost > ast.cost + 1e-9  # the detour is genuinely longer than omniscient A*


def test_ad_star_no_path_when_walled_off() -> None:
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."])
    res = ADStar(config("ad_star")).plan(grid, (0, 0), (0, 4))
    assert not res.success
    assert res.path == []
    assert res.cost == 0.0


def test_ad_star_rejects_out_of_range_param(tmp_path: Path) -> None:
    # eps_start declares min 1.0; a below-min default must fail at load time so an
    # inflation factor < 1 (which would break the suboptimality bound) never runs.
    cfg = write_config(
        tmp_path / "bad_ad.yaml",
        "ad_star",
        [{"name": "eps_start", "type": "float", "default": 0.5,
          "min": 1.0, "max": 10.0, "description": "below min"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(cfg)


def test_ad_star_anytime_solutions_improve_to_optimum(tmp_path: Path) -> None:
    grid = grid_from(_FIELD_ROWS)
    # Large sensor radius: the whole field is sensed from the start, so belief == truth
    # and no motion-time changes muddy the ε sweep — the anytime improvement is
    # isolated to the initial position, exactly like a static ARA* sweep.
    cfg = _anytime_config(tmp_path, eps_start=4.0, eps_step=0.5, sensor_radius=50)
    buf = io.StringIO()
    res = ADStar(cfg).plan(grid, _FIELD_START, _FIELD_GOAL, TraceRecorder(buf))

    events = [json.loads(line) for line in buf.getvalue().splitlines()]
    # Collect the plans published while the robot is still at the start (before it
    # leaves on its first move, i.e. before the 2nd robot_moved event).
    moves = 0
    start_phase: list[list[Cell]] = []
    for e in events:
        if e["event"] == "robot_moved":
            moves += 1
        if e["event"] == "path_found" and moves <= 1:
            start_phase.append([tuple(s) for s in e["path"]])
    costs = [_path_cost(grid, p) for p in start_phase]

    assert len(costs) >= 2  # genuinely anytime: at least one repair after the first
    assert costs == sorted(costs, reverse=True)  # each solution no worse than the next
    assert costs[0] > costs[-1]  # the first ε-inflated plan is strictly suboptimal here
    ast = AStar(config("astar")).plan(grid, _FIELD_START, _FIELD_GOAL)
    assert costs[-1] == pytest.approx(ast.cost)  # final ε -> 1 plan is the optimum
    assert res.cost == pytest.approx(ast.cost)  # executed trajectory is optimal too

File: cpp/tests/test_ad_star.cpp
#include <sstream>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/search/ad_star.hpp"
#include "navigation/global_planning/search/astar.hpp"
#include "navigation/maps/loader.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Cell;

namespace {

core::ParamSet cfg(const std::string& algo) {
  return core::ParamSet::from_yaml(test::repo_path("configs/global_planning/" + algo + ".yaml"));
}

// Writes an AD* config with an explicit schedule + sensor radius so tests can force a
// specific ε sweep / sensing horizon without depending on the shipped defaults.
core::ParamSet ad_cfg(double eps_start, double eps_final, double eps_step, int sensor_radius) {
  std::ostringstream y;
  y << "algorithm: ad_star\ncategory: global_planning\nparams:\n"
    << "  - name: eps_start\n    type: float\n    default: " << eps_start
    << "\n    min: 1.0\n    max: 10.0\n    description: start\n"
    << "  - name: eps_final\n    type: float\n    default: " << eps_final
    << "\n    min: 1.0\n    max: 10.0\n    description: final\n"
    << "  - name: eps_step\n    type: float\n    default: " << eps_step
    << "\n    min: 0.01\n    max: 10.0\n    description: step\n"
    << "  - name: sensor_radius\n    type: int\n    default: " << sensor_radius
    << "\n    min: 1\n    max: 50\n    description: radius\n"
    << "  - name: max_expansions\n    type: int\n    default: 100000\n    min: 1\n"
    << "    max: 100000000\n    description: cap\n";
  return core::ParamSet::from_yaml(test::write_temp("ad_star.yaml", y.str()));
}

maps::OccupancyGrid2D& as_grid(core::MapBase& m) {
  return *dynamic_cast<maps::OccupancyGrid2D*>(&m);
}

// Every consecutive pair is a legal move under GROUND TRUTH (grid.neighbors), proving
// the executed trajectory never crosses a real obstacle.
bool valid_trajectory(const maps::OccupancyGrid2D& grid, const std::vector<Cell>& path) {
  for (size_t i = 0; i + 1 < path.size(); ++i) {
    bool ok = false;
    for (const auto& [c, w] : grid.neighbors(path[i])) {
      (void)w;
      if (c == path[i + 1]) {
        ok = true;
        break;
      }
    }
    if (!ok) return false;
  }
  return true;
}

// Same field as the ARA* anytime fixture but start/goal swapped so the two obstacles
// sit near the robot: AD*'s backward search rushes toward the robot under the inflated
// key and rounds them the costly way first, then repairs to the optimum.
const std::vector<std::string> kField = {
    ".............", "..........#..", "...........#.", ".............", ".............",
    ".............", ".............", ".............", ".............", ".............",
    ".............", ".............", ".............", "............."};
const Cell kFieldStart{0, 11};
const Cell kFieldGoal{11, 12};

}  // namespace

// (a) static map: the executed trajectory equals the omniscient A* optimum -----------
TEST(AdStar, StaticTrajectoryMatchesAstarOptimumOnMaze01) {
  auto map = maps::load_map(test::repo_path("maps/grid/maze01.yaml"), 0, 8);
  auto& grid = as_grid(*map);
  auto sc = maps::load_scenario(test::repo_path("maps/scenarios/maze01_s1.yaml"));
  Cell start = grid.world_to_cell(sc.start.x, sc.start.y);
  Cell goal = grid.world_to_cell(sc.goal.x, sc.goal.y);

  global_planning::AdStarPlanner ad(cfg("ad_star"));
  auto ra = ad.plan(grid, start, goal, nullptr);
  global_planning::AstarPlanner astar(cfg("astar"));
  auto ro = astar.plan(grid, start, goal, nullptr);

  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_TRUE(valid_trajectory(grid, ra.path));
  EXPECT_NEAR(ra.cost, ro.cost, 1e-9);
}

// (b) replan: a valid, strictly longer detour once the C-trap's back wall is sensed ---
TEST(AdStar, ReplansValidDetourWhenTrapRevealed) {
  auto map = maps::load_map(test::repo_path("maps/grid/dstar_trap01.yaml"), 0, 8);
  auto& grid = as_grid(*map);
  auto sc = maps::load_scenario(test::repo_path("maps/scenarios/dstar_trap01_s1.yaml"));
  Cell start = grid.world_to_cell(sc.start.x, sc.start.y);
  Cell goal = grid.world_to_cell(sc.goal.x, sc.goal.y);

  global_planning::AdStarPlanner ad(cfg("ad_star"));
  auto ra = ad.plan(grid, start, goal, nullptr);
  global_planning::AstarPlanner astar(cfg("astar"));
  auto ro = astar.plan(grid, start, goal, nullptr);

  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_TRUE(valid_trajectory(grid, ra.path));
  EXPECT_GE(ra.stats.iterations, 1);   // replanned after sensing the trap
  EXPECT_GT(ra.cost, ro.cost + 1e-9);  // the detour is genuinely longer than omniscient A*
}

// (c) no-path case -------------------------------------------------------------------
TEST(AdStar, NoPathWhenWalledOff) {
  auto g = test::make_grid({"..#..", "..#..", "..#..", "..#..", "..#.."});
  global_planning::AdStarPlanner ad(cfg("ad_star"));
  auto r = ad.plan(g, Cell{0, 0}, Cell{0, 4}, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
  EXPECT_EQ(r.cost, 0.0);
}

// (d) param validation failure -------------------------------------------------------
TEST(AdStar, BadEpsStartThrows) {
  std::string bad = test::write_temp(
      "ad_star.yaml",
      "algorithm: ad_star\ncategory: global_planning\nparams:\n"
      "  - name: eps_start\n    type: float\n    default: 0.5\n    min: 1.0\n    max: 10.0\n"
      "    description: below min\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

// (e) anytime: an inflated first trajectory, then repair to the optimum. Large sensor
// radius makes belief == truth from the start, isolating the ε sweep to a static
// improvement so the first (ε fixed high) executed trajectory is strictly suboptimal.
TEST(AdStar, AnytimeSolutionsImproveToOptimum) {
  auto grid = test::make_grid(kField);

  // First solution only: ε pinned at 4 (eps_final = 4), so the robot moves on a
  // suboptimal frozen g.
  global_planning::AdStarPlanner first(ad_cfg(4.0, 4.0, 1.0, 50));
  auto rf = first.plan(grid, kFieldStart, kFieldGoal, nullptr);
  ASSERT_TRUE(rf.success);

  // Full sweep ε 4 -> 1, capturing the emitted anytime solutions.
  std::ostringstream os;
  core::TraceRecorder rec(os);
  global_planning::AdStarPlanner ad(ad_cfg(4.0, 1.0, 0.5, 50));
  auto ra = ad.plan(grid, kFieldStart, kFieldGoal, &rec);
  ASSERT_TRUE(ra.success);

  int path_found = 0;
  std::istringstream in(os.str());
  for (std::string line; std::getline(in, line);) {
    if (line.find("\"event\":\"path_found\"") != std::string::npos) ++path_found;
  }
  EXPECT_GE(path_found, 2);  // genuinely anytime: at least one repair after the first

  global_planning::AstarPlanner astar(cfg("astar"));
  auto ro = astar.plan(grid, kFieldStart, kFieldGoal, nullptr);
  EXPECT_NEAR(ra.cost, ro.cost, 1e-9);  // converged to the optimum
  EXPECT_GT(rf.cost, ra.cost);          // first ε-inflated trajectory is strictly suboptimal
}
